### PR TITLE
Transition Discovery into a stage

### PR DESCRIPTION
**Pull request by jeremyfowers**
# Note to reviewers

A large amount of cleanup tasks are already documented in #168 

# Summary of Changes

- Discovery is now a stage called `discover`, which takes one input python script, discovers the models within, and passes one model/inputs to the next Stage.
- **New feature**: all calls to turnkey have a stats.yaml and state.yaml file now, even if discovery fails.
- **New feature**: `discover` is no longer a mandatory way to start the sequence.
  - `onnx-load` is now an alternative way to start the sequence, which does not invoke the discovery code path whatsoever.
  - In the future we can add more entry points! Load-build-from-cache is a good one. We can also add bespoke entry points for specific model types (e.g., LLM), rather than trying to handle everything generically as a pytorch model or onnx model.
- A lot of functionality has moved out of the `turnkeyml.analysis` module
  - `turnkeyml.analysis` is now only responsible for discovering models. It no longer does error handling, stats, nor does it call `build_model()`
  - The main API now calls `build_model()` and saves the high-level stats
  - `Sequence` is responsible for all error handling.
- Only one model is passed from `discover` to subsequent stages.
  - By default, it is the first model in the input script.
  - Users can specify a model with the `::HASH` postfix, as before, however now only one HASH is allowed.
  - This allows us to change the naming scheme for builds:
    - `turnkey -i SCRIPT.py discover` will name the build `SCRIPT`
    - `turnkey -i SCRIPT.py::HASH discover` will name the build `SCRIPT_HASH`

- The concept of an `evaluation_id` has been eliminated.
  - `stats.yaml` is flat now; there is no `evaluations` section.
  - The stats-related APIs have changed to simply `save_stat()`, instead of `save_model_stat()` and `save_model_eval_stat()` 
- The concept of a `ModelType` enum has been eliminated. 
  - Code that cares about the type of the model should just test the type of the model in-situ instead of relying on global model type detection.

### CODE REVIEW

File: .github/workflows/test_turnkey.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
     steps:
       - uses: actions/checkout@v3
       - name: Set up Miniconda with 64-bit Python
-        uses: conda-incubator/setup-miniconda@v2
+        uses: conda-incubator/setup-miniconda@v3
```
Comment (danielholanda) [inline .github/workflows/test_turnkey.yml:28]: Any particular reason we are upgrading this?
There is nothing wrong with it, I'm just curious.

File: src/turnkeyml/analyze/discover.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                        model_info.params,
+                    )
+
+                    state.save_stat(
+                        fs.Keys.CLASS,
+                        type(model_info.model).__name__,
+                    )
+
+                    state.results = model_info.model
+                    # FIXME: we should be able to get rid of state.model now
```
Comment (danielholanda) [inline src/turnkeyml/analyze/discover.py:154]: Recommend to either fix or add an issue here.

**Review by danielholanda — APPROVED** [review]
Approved with suggestions.

File: .github/workflows/test_turnkey.yml (hunk 1 above)
Comment (jeremyfowers) [reply .github/workflows/test_turnkey.yml:28]: There was some crazy livesite last week: https://github.com/conda-incubator/setup-miniconda/issues/358

Ultimately, I'm not sure if this upgrade helped, but I did it as a precaution and it definitely didn't hurt.

File: src/turnkeyml/analyze/discover.py (hunk 2 above)
Comment (jeremyfowers) [reply src/turnkeyml/analyze/discover.py:154]: Documented in #168

File: test/cli.py [hunk 3]
```diff
@@ -857,7 +866,6 @@ def test_024_input_text_file(self):
             builds_found == 3
         ), f"Expected 3 builds (1 for linear.py, 2 for linear2.py), but got {builds_found}."
 
-    @unittest.skip("Flaky test https://github.com/onnx/turnkeyml/issues/58")
```
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply test/cli.py:860]: AFAIK, yes! The big problem with the prior test is that timing out during discovery would result in a fail, so the timeout had to be long enough to get past discovery, but not so long that the build succeeded. Now timing out during discovery yields a useful stats file, so the test is much less flakey.

File: .github/workflows/publish-to-test-pypi.yml [hunk 4]
```diff
@@ -15,7 +15,7 @@ jobs:
     runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply .github/workflows/publish-to-test-pypi.yml:15]: Reproduced, definitely a bug!

File: .github/workflows/publish-to-test-pypi.yml [hunk 5]
```diff
@@ -15,7 +15,7 @@ jobs:
...
     steps:
```
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply .github/workflows/publish-to-test-pypi.yml:16]: reproduced with `turnkeyml\examples\cli\onnx> turnkey -i .\hello_world.onnx onnx-load benchmark`

Definitely a bug!

File: .github/workflows/publish-to-test-pypi.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply .github/workflows/publish-to-test-pypi.yml:15]: The `select with `-i _builder.py::465f6391` is also wrong, they should also say `-i resnet18.py::HASH`

**Review by ramkrishna2910 — APPROVED** [review]

**Comment by ramkrishna2910** [comment]
All cache folders include `_turnkey` in the end
image
Is that necessary?

File: .github/workflows/publish-to-test-pypi.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ramkrishna2910) [reply .github/workflows/publish-to-test-pypi.yml:16]: Reproduced this

File: .github/workflows/publish-to-test-pypi.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply .github/workflows/publish-to-test-pypi.yml:15]: Fixed both in https://github.com/onnx/turnkeyml/pull/174/commits/4cbbf3e5daff2c3ca5078248b5957c9d38961715

File: src/turnkeyml/analyze/discover.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        parsed_args.input = file_path
+
+        if len(targets) > 1:
+            raise exp.ArgError(
+                "Only one target (number after the ::) is allowed, "
+                f"but received {encoded_input}"
+            )
+        elif len(targets) == 1:
+            parsed_args.target = targets[0]
+        else:  # len(targets)==0
```
Comment (danielholanda) [inline src/turnkeyml/analyze/discover.py:79]:
```suggestion
        else:
```

**Review by danielholanda — APPROVED** [review]

File: .github/workflows/publish-to-test-pypi.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply .github/workflows/publish-to-test-pypi.yml:16]: image

fixed in https://github.com/onnx/turnkeyml/pull/174/commits/b9e3771772682e9454f696c4accad9a5b59381bf

File: src/turnkeyml/build/export.py [hunk 7]
```diff
@@ -85,9 +86,7 @@ class OnnxLoad(stage.Stage):
     Stage that takes an ONNX model as input and passes it to the following
     stages.
 
-    Expected inputs:
-     - state.model is a path to the ONNX model
-     - state.inputs is a dict that represents valid inputs for the onnx model
+    Expected inputs: None
 
     Outputs:
      - A *-base.onnx file that implements state.model given state.inputs.
```
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply src/turnkeyml/build/export.py]: fixed in https://github.com/onnx/turnkeyml/pull/174/commits/5ff7465486521220526ec8ca2fce528fd5179338

File: src/turnkeyml/build/export.py [hunk 8]
```diff
@@ -148,22 +148,23 @@ def fire(self, state: fs.State):
                     """
                     raise exp.StageError(msg)
 
+        print(opset)
```
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply src/turnkeyml/build/export.py]: newp https://github.com/onnx/turnkeyml/pull/174/commits/2a9ebbc61aaaf7da1468e44a40e418744e112a61

**Comment by jeremyfowers** [comment]
> All cache folder names include `_turnkey` in the end image Is that necessary?

@ramkrishna2910 its not all models. The `examples/cli/scripts` models all have a label `author::turnkey` and the author name is auto-populated into the build dir

File: src/turnkeyml/analyze/discover.py (hunk 6 above)
Comment (jeremyfowers) [reply src/turnkeyml/analyze/discover.py:79]: Wont fix: i find it helpful to label else statements if they have a specific intent. that way if the if or elif statements change, you know if you need to change the else to sustain the intent.

**Pull request merged**